src/python/WMCore/MicroService/MSPileup/MSPileupData.py

```diff
@@ -79,6 +79,36 @@ def customDID(pname):
     return pname
 
 
+def updateTransitionRecord(doc, userDN, logger):
+    """
+    Update transition record based on the following logic:
+    - check previous fraction value and add new transition record if it is
+    - differ from current one or does not exist
+      - first use-case: differ from container fraction of record
+      - second use-case: differ from last transition container fraction record
+    - https://gist.github.com/amaltaro/b4f9bafc0b58c10092a0735c635538b5
+
+    :param doc: MSPileup record
+    :param userDN: user DN (string)
+    :param logger: logger to use
+    :return: nothing (document transition is updated in place)
+    """
+    fraction = doc['containerFraction']
+    transition = doc.get('transition', [])
+    prevTranRecord = transition[-1]
+    userDN = userDN or prevTranRecord['DN']
+    if prevTranRecord['containerFraction'] != fraction:
+        customName = customDID(prevTranRecord['customDID'])
+        # preserve previous container fraction
+        transitionRecord = {'containerFraction': fraction,
```

Comment:
We should only set it to the container fraction IF we have succeeded resizing the pileup container and making all the data placement logic for the custom container. 

In case we hit an exception while executing the partial pileup logic, this different containerFraction is the only hint that we have to re-execute the same logic in the next cycle. In other words, containerFraction == transition/containerFraction tells the service that nothing needs to be done in terms of partial pileup.

Reply:
> But in this comment you write that we should update if we have succeeded resizing which ... part of upstream code logic

Yes, I fully agree on that. Hence my concern here.

Bottom line, this method is fairly generic so far and it should not do anything other than:
1) validating the document content (if required - or we can refactor it in the future with the proper separation)
2) persisting the document in the backend database.

Meaning, there should be no key/value addition and or assignment in here.

Reply:
> ok, that implies that update of transition records should be done only by MSPileupTasks

Update of the transition record, yes. But not the creation of a record.

Reviewing the gist, we create a new transition record in two scenarios:
* upon creation of a new pileup document
* upon update of the containerFraction
and we should keep it like that, as we have already exhaustively discussed it.

Now thinking about the interactions with `updatePileup`, we have:
1) pileup document update through MSPileup (REST), without changing `containerFraction`. The method should only persist the document - and validate, if required.
2) pileup document update through MSPileup (REST), with changes to `containerFraction`. A transition record **must** be created **somewhere** between the REST/Data layer. Then the document is persisted - and validated, if required.
3) pileup document update through MSPileupTasks. Regardless of `containerFraction` changes or not, it should only persist the document - and validate, if required.

Am I missing any other interactions? Maybe what we need is to have a hook such that we can identify when a document is provided by the end user or by the service daemon (maybe a method argument).